### PR TITLE
feat: added support for an empty string

### PR DESCRIPTION
**Pull request by PMAWorks**
Adds support for empty strings through the flag "allowEmptyString".

There are 2 implementation options:
1. Through token parsing and map comparison
2. Through the insert `&nbsp;`

In the end, I want to leave it `&nbsp;`, but maybe there will be ideas for the second option

### CODE REVIEW

**Comment by makhnatkin** [comment]
I propose incorporating this functionality into the markup mode.

1. Implement auto-completion software and provide additional auto-completion features.
2. Assign a keyboard shortcut.

File: src/markup/commands/emptyRow.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+                range.head + str.length,
+                range.goalColumn,
+                range.bidiLevel ?? undefined,
+            ),
+        };
+    });
+
+    dispatch(state.update(trSpec));
+    return true;
+};
```
Comment (d3m1d0v) [inline src/markup/commands/emptyRow.ts:22]: I think you can use `replaceOrInsertAfter` helper, like in `insertHRule` command

File: src/markup/commands/helpers.ts [hunk 2]
```diff
@@ -39,14 +43,26 @@ export function replaceOrInsertAfter(state: EditorState, markup: string): Transa
                 state.lineBreak.repeat(extraBreaks.after),
         );
     } else {
-        const insert = state.lineBreak.repeat(2) + markup + state.lineBreak.repeat(2);
+        let breaksAfter = 2;
+
+        if (shouldGetBreaks) {
```
Comment (makhnatkin) [inline src/markup/commands/helpers.ts:48]: use from `emptyRows`

File: src/markup/codemirror/yfm.ts [hunk 3]
```diff
@@ -64,23 +60,6 @@ const MarkedExtension = mdInlineFactory({
     tag: customTags.marked,
 });
 
-export type YfmNoteType = 'info' | 'tip' | 'warning' | 'alert';
```
Comment (makhnatkin) [inline src/markup/codemirror/yfm.ts:67]: return plz

File: src/extensions/base/BaseSchema/BaseSchemaSpecs/index.ts [hunk 4]
```diff
@@ -62,9 +63,16 @@ export const BaseSchemaSpecs: ExtensionAuto<BaseSchemaSpecsOptions> = (builder,
                     : undefined,
             },
             fromMd: {tokenSpec: {name: BaseNode.Paragraph, type: 'block'}},
-            toMd: (state, node) => {
-                state.renderInline(node);
-                state.closeBlock(node);
+            toMd: (state, node, parent) => {
+                if (opts.allowEmptyRows && !node.content.size) {
+                    const isParentEmpty = parent.content.size / parent.content.childCount === 2;
```
Comment (makhnatkin) [inline src/extensions/base/BaseSchema/BaseSchemaSpecs/index.ts:68]: add comment plz

Comment (makhnatkin) [reply src/extensions/base/BaseSchema/BaseSchemaSpecs/index.ts:68]: There are nodes (for example, iframe) that contain size === 1, we suggest changing the logic to bypass the nodes

File: src/bundle/useMarkdownEditor.ts [hunk 5]
```diff
@@ -82,6 +89,7 @@ export function useMarkdownEditor<T extends object = {}>(
                 html: md.html ?? props.allowHTML,
                 linkify: md.linkify ?? props.linkify,
                 linkifyTlds: md.linkifyTlds ?? props.linkifyTlds,
+                pmTransformers: pmTransformers,
```
Comment (makhnatkin) [inline src/bundle/useMarkdownEditor.ts:92]: `pmTransformers,`

File: src/core/markdown/ProseMirrorTransformer/getTransformers.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    emptyRowTransformer?: boolean;
+};
+
+type GetPMTransformersType = (config: GetTransformersProps) => TransformFn[];
+
+export const getPMTransformers: GetPMTransformersType = ({emptyRowTransformer}) => {
+    const transformers = [];
+
+    if (emptyRowTransformer) {
+        transformers.push(transformEmptyParagraph);
```
Comment (makhnatkin) [inline src/core/markdown/ProseMirrorTransformer/getTransformers.ts:15]: Add today: add a new method to the ExtensionBuilder

File: demo/components/Playground.tsx [hunk 7]
```diff
@@ -77,6 +77,7 @@ export type PlaygroundProps = {
     allowHTML?: boolean;
     settingsVisible?: boolean;
     initialEditor?: MarkdownEditorMode;
+    preserveEmptyRows?: boolean;
```
Comment (makhnatkin) [inline demo/components/Playground.tsx:80]: Add preserveEmptyRows to the following [line](https://github.com/gravity-ui/markdown-editor/blob/b536dc054790900eafd3997a8d7e88fdd396163a/demo/components/Playground.tsx#L239) in Playground.tsx to enable rerendering in the demo.

**Review by makhnatkin — APPROVED** [review]

File: src/bundle/Editor.ts [hunk 8]
```diff
@@ -279,7 +281,11 @@ export class EditorImpl extends SafeEventEmitter<EventMapInt> implements EditorI
                     extensions: this.#markupConfig.extensions,
                     disabledExtensions: this.#markupConfig.disabledExtensions,
                     keymaps: this.#markupConfig.keymaps,
-                    yfmLangOptions: {languageData: this.#markupConfig.languageData},
+                    yfmLangOptions: {
+                        languageData: getAutocompleteConfig({
+                            preserveEmptyRows: this.#mdOptions.preserveEmptyRows,
+                        }),
+                    },
```
Comment (d3m1d0v) [inline src/bundle/Editor.ts:288]: you should concat languageData config from `this.#markupConfig.languageData` and result of call `getAutocompleteConfig()` together

File: src/bundle/types.ts [hunk 9]
```diff
@@ -39,9 +41,11 @@ export type WysiwygPlaceholderOptions = {
 
 export type MarkdownEditorMdOptions = {
     html?: boolean;
+    preserveEmptyRows?: boolean;
     breaks?: boolean;
     linkify?: boolean;
     linkifyTlds?: string | string[];
+    pmTransformers?: TransformFn[];
```
Comment (d3m1d0v) [inline src/bundle/types.ts:48]: is this field not needed anymore?

File: src/bundle/types.ts [hunk 10]
```diff
@@ -39,9 +41,11 @@ export type WysiwygPlaceholderOptions = {
 
 export type MarkdownEditorMdOptions = {
     html?: boolean;
+    preserveEmptyRows?: boolean;
```
Comment (d3m1d0v) [inline src/bundle/types.ts:44]: move this setting to experiment options

File: src/bundle/config/action-names.ts [hunk 11]
```diff
@@ -20,6 +20,23 @@ const names = [
...
+    'quote',
+    'yfm_cut',
+    'yfm_note',
+    'yfm_block',
+    'yfm_html_block',
+    'yfm_layout',
+    'table',
+    'code_inline',
+    'code_block',
+    'image',
```
Comment (d3m1d0v) [inline src/bundle/config/action-names.ts:39]: duplicated?

File: src/core/markdown/ProseMirrorTransformer/getTransformers.ts (hunk 6 above)
Comment (d3m1d0v) [reply src/core/markdown/ProseMirrorTransformer/getTransformers.ts:16]: > Add today: add a new method to the ExtensionBuilder

in the future

**Review by d3m1d0v — APPROVED** [review]

**Pull request merged**